Use etld plus one matching for 3p

snyderp opened the pull request:
fixes #171

diracdeltas commented on test/js/matchingTest.js:
```diff
@@ -455,4 +455,35 @@ describe('matching', function () {
+        'www.example.org'
+      )
+      assert.equal(queryResult.matches, true)
+    })
```
i suggest adding some tests that use non-trivial etld+1's like, `example.co.uk` and `example.githubusercontent.com`

snyderp replied:
There are many, but there at the C++ level.  They're in this file https://github.com/brave/ad-block/blob/use-etld-plus-one-matching-for-3p/test/etld_test.cc

snyderp replied:
ex: https://github.com/brave/ad-block/blob/use-etld-plus-one-matching-for-3p/test/etld_test.cc#L488

diracdeltas replied:
i see thanks

diracdeltas commented:
general question about this approach:

is it possible to just use the etld+1 parsing from Chromium? see https://cs.chromium.org/chromium/src/net/base/registry_controlled_domains/registry_controlled_domain.h?q=getdomainandregistry&dr=CSs

snyderp commented:
> general question about this approach:
> 
> is it possible to just use the etld+1 parsing from Chromium? see https://cs.chromium.org/chromium/src/net/base/registry_controlled_domains/registry_controlled_domain.h?q=getdomainandregistry&dr=CSs

We could, but then we'd loose the ability to run in node (which has been very valuable for crawling / measurement, getting other folks to use the code, debugging, etc)

fmarier commented on etld/domain.cc:
```diff
@@ -0,0 +1,51 @@
+/* Copyright (c) 2018 The Brave Software Team. Distributed under the MPL2
```
nit: Should be 2019. This applies to all of the new files added in this PR.

snyderp replied:
fixed :)

fmarier replied:
Maybe I'm confused about how GitHub is displaying the latest version of your PR, but it seems like you fixed all files, except for `etld/domain.cc`?

snyderp replied:
No, i goofed, apologies for the butter fingers.  Fixed now

fmarier commented:
I've got a question about the build process since I don't actually know when the build step takes place: if we pull down the list at build time, is there a reason to have it checked into the repo?

snyderp commented:
> I've got a question about the build process since I don't actually know when the build step takes place: if we pull down the list at build time, is there a reason to have it checked into the repo?

You're right no need for this.  I removed it from the .gitignore previously, now also removed it from the set of tracked files.  Should be good now

snyderp commented:
@bbondy my code expects the public suffix list to be in a known location, and lazily parses the list on first use (i.e. at `etld/data/<list>.dat`).  I have no idea if this will work when rolled into the larger browser.  I'm just not familiar enough with the build process.  Can you double check that aspect?

fmarier commented on .gitignore:
```diff
@@ -48,3 +48,9 @@ out
+.vscode
+
+# These files are either fetched at build time, or generated from the build
+etld/data/public_suffix_list.h
```
nit: I usually try to make these absolute paths with respect to the location of the `.gitignore` file (i.e. `/etld/data/public_suffix.h`) in order to avoid unexpected matches somewhere else in the codebase. Not very likely in this case, so feel free to ignore this suggestion, but a good habit IMHO.

fmarier commented on Makefile:
```diff
@@ -5,7 +5,9 @@
 
 build:
-	 ./node_modules/.bin/node-gyp configure && ./node_modules/.bin/node-gyp build
+	curl -s https://publicsuffix.org/list/public_suffix_list.dat -o etld/data/public_suffix_list.dat
```
Kudos for not using `-k` here :)

nit: If you want to make that `curl` line even stricter, you could also throw in `--tlsv1.2` to enforce a minimum level of TLS.

fmarier commented on etld/data/build.js:
```diff
@@ -0,0 +1,73 @@
+  constructorArgs.push(isException ? "true" : "false");
+
+  const wrappedLabels = labels.map(JSON.stringify);
+  constructorArgs.push("{" + wrappedLabels.join(", ") + "}");
```
Should the labels be surrounded by `"` in case they're not all bare words? Also, is it possible they include `"` characters that should be escaped?

fmarier commented on etld/domain.cc:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  if (previous != 0) {
+    labels_.push_back(string.substr(previous, current - previous));
```
I believe this call will fail if you get invalid input like: `abcd.efgh.` (note the trailing dot).

snyderp commented:
@bbondy this is now ready for review again. The ways to enable the eTLD+1 checking (by parsing a public suffix list) are:

1) when using the `check.js` script, use the new `-P, --public-suffix-rules-path` option, and point it to a text including public suffix rules.
2) use the js `AdBlockClient.parsePublicSuffixRules` method and give it a string containing public suffix rules
3) use the C++ `AdBlockClient::parsePublicSuffixRules` method with a `char*` / `std::string` of rules
4) use either the C++ or JS `deserialize` methods with a `dat` file that includes public suffix rules data (serializing after doing 1, 2 or 3 will include the public suffix rule data in the `.dat`).
